src/main/java/org/eclipse/yasson/YassonConfig.java

```diff
@@ -45,7 +45,24 @@
      * @see #withEagerParsing(Class...)
      */
     public static final String EAGER_PARSE_CLASSES = "yasson.eager-parse-classes";
-    
+
+    /**
+     * @see #withMapToObjectSerializerNullKey(String)
+     */
+    public static final String MAP_OBJECT_SERIALIZER_NULL_KEY = "yasson.mapToObjectSerializer.nullKey";
```

Comment:
All the maps (no matter the key type) can have a null in the key. That key should be represented by a string value (now it's "null" at serializing but it's never read in the deserialization part, the string is just read like that, "null" string not the null value, so the null value is lost right now). The idea was having a property you can change the key value for null (and yes, globally). By default it is "null" as now. It can be something like "yasson.null" which is more complicated to be a valid String key (a real key which is that string after serialization). I thought that a global conf was enough but if you see a better idea just let me know.

But, as I commented in the description, this is an extra to allow the null key to be configurable (use a different string in case it collides with a real key). We can just remove this part and use always "null" or other fixed value.

I'll re-do the PR with all the changes when we agree what to do with this point. :-)

Thanks a lot for looking this!

Reply:
Oh I miss looked here. it is serializer stuff I was actually talking about deserialization. Sorry. But anyway we should be able to get the same Map from the json as we were using to create it. So even though we will use the substitute token to replace null value key, it is a bit unclear to me how it would behave in terms of deserialization (to obtain the same Map as we had before) :-) I will try to think about it a bit, but so far I do net exactly see any clear solution.

Reply:
Yes, exactly. The property is used in both: the serializer, when null is detected the string used as key is obtained from that property; and the deserializer, when the key read is exactly the property value it is considered the null key. The property is just added to let the user define a new value if it collides with a real key (as in json the key is a string, hypothetically it can always collide with a real key). And the same, I used default "null" value because it was used before, but we can choose a very weird default string to minimize the problem ("org.eclipse.yasson.null-key" for example).

Reply:
@Verdent Maybe we are missing the obvious solution. The problem is that the MapToObjectSerializer (`{"key1":"value1",...}`) cannot represent null keys. Just because JSON doesn't allow null keys (`{null:"value1"}`is not valid), only strings, but java allows them. So [this comment](https://github.com/eclipse-ee4j/yasson/blob/2.0.2/src/main/java/org/eclipse/yasson/internal/serializer/MapSerializer.java#L129-L130) in the serializer is wrong.

What if we don't use this impl if there is a null key in the map and use the MapToEntriesArraySerializer one? This avoids all the property stuff, and, in the end, null keys are not managed OK right now, so we are breaking anything that was not broken before.

Reply:
I have added my proposal, take a look.

Reply:
@rmartinc I kind of like the idea of doing this the way you proposed now. The only problem is, that it is backwards incompatible change. What we should probably do is to allow this enhancement via implementation specific config property. Default behavior for this null handling would as is right now, but if anyone knows what he is doing, I think that having a way to change the null handling is fine.

Reply:
@Verdent OK, perfect, I'll add a new property as I did previously with my first PR but using a boolean prop to revert the old behavior. I'll update the PR probably on Monday. :-)

Reply:
Done!

A new property `yasson.map-object-serializer-for-null-keys` was defined (default value is false). I also added a new test in `SerializersTest.java` called `testSerializeMapWithNullsBackwardsCompatible` which tests that the same output is got when serializing a Map with a null key and the prop is set to true (previous behavior).

Reply:
Thank you for the change, but the problem is, that the "new" behavior has to be enabled via property and not used by default. :-)

Reply:
Argh! I understood the opposite, OK, it's modified now. Changed the property name to `yasson.force-map-array-serializer-for-null-keys` and the tests now expect the new output only when that prop is set.

Just a note, right now a null key would be transformed to String "null" if the `MapToObjectSreializer` is chosen, which mainly means that the serialization/deserialization for that map is broken. That's probably why I understood that I had to set the new behavior as the default one. But as nulls as map keys are rare (and a bad practice) I don't think this will be an important problem. The new property can be set to make it work.